firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/common/NativeSessionFile.java

```diff
@@ -0,0 +1,40 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.crashlytics.internal.common;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import com.google.firebase.crashlytics.internal.model.CrashlyticsReport;
+import java.io.InputStream;
+
+/**
+ * An abstraction of a "File" resource sent to the Crashlytics backend when an NDK Crash has
+ * occurred.
+ */
+interface NativeSessionFile {
+  /** Shortname of the file, e.g., "logs" */
+  @NonNull
+  String getName();
+
+  /** Representation of the NativeSessionFile as a stream */
+  @Nullable
+  InputStream getStream();
+
+  /**
+   * Representation of the NativeSessionFile as {@link CrashlyticsReport.FilesPayload.File} object.
+   */
+  @Nullable
+  CrashlyticsReport.FilesPayload.File asFilePayload();
```

Comment:
Short answer: Eventually getStream() will go away because this class will not have the old codepath, and this will just **be** the FilesPayloadSerializer.

Long answer:  
Yeah, I went back and forth on this a little bit myself. I didn't want to build the datatransport code on getStream() because it results in a double copy of a byte stream into a new byte stream.

(wearing OOP nerd hat) There's actually a way to solve this where you separate out the data (bytes, files) and behavior (gzip from stream of bytes, gzip from filestream, datatransport from bytes of files or bytes of bytes).

The way is multiple dispatch. On java though, multiple dispatch has to be implemented via visitor pattern (implement a stream converter that visits the bytes / files, then converts to a stream, implement a filepayloadconverter that visits the bytes / files, and converts them to bytes). Doing so is pretty verbose. Since the API for getStream() is going away, I decided it was okay to just flatten them into this api, but if you wanted, I can switch to visitor.

This whole multiple dispatch challenge, for what it is worth, is called the "expression problem" https://eli.thegreenplace.net/2016/the-expression-problem-and-its-solutions/

Reply:
🤓 alright, leave it the way it is. 😝